source/open_node/ast.py

```diff
@@ -0,0 +1,1017 @@
+    """
+
+    statements: List[Statement]
+    version: dict[str, int] = field(init=False, default_factory=dict, compare=False)
+    io_variables: List[IODeclaration] = field(init=False, default_factory=list)
+
+
+class Statement(OpenNode):
+    """A statement: anything that can appear on its own line"""
+
+
+@dataclass
+class AliasStatement(Statement):
+    """
+    Alias statement
+
+    Example::
+
+        let a = qubits[0];
+
+    """
+
+    target: Identifier
+    value: Union[IndexIdentifier, Identifier]
+
+
+@dataclass
+class ExpressionStatement(Statement):
+    """A statement that contains a single expression"""
+
+    expression: Expression
+
+
+class QubitDeclTypeName(Enum):
+    qubit = auto()  # OpenQASM 3 style.
+    qreg = auto()  # OpenQASM 2 style.
+
+
+@dataclass
+class QubitDeclaration(Statement):
+    """
+    Global qubit declaration
+
+    Example::
+
+        qubit q;
+        qubit[4] q;
+
+        q // <- qubit
+        [4] // <- designator
+
+    """
+
+    decl_type_name: QubitDeclTypeName
+    qubit: Qubit
+    designator: Optional[Expression]
+
+
+@dataclass
+class QuantumGateDefinition(Statement):
+    """
+    Define a new quantum gate
+
+    Example::
+
+        gate cx c, t {
+            ctrl @ unitary(pi, 0, pi) c, t;
+        }
+
+    """
+
+    name: str
+    arguments: List[ClassicalArgument]
+    qubits: List[Identifier]
+    body: List[QuantumStatement]
+
+
+class QuantumStatement(Statement):
+    """Statements that may appear inside a gate declaration"""
+
+
+@dataclass
+class QuantumForInLoop(Statement):
+    """For In loops that contain only quantum statements."""
+
+    loop_variable: Identifier
+    set_declaration: Union[RangeDefinition, List[Expression], Identifier]
+    block: List[QuantumStatement]
+
+
+@dataclass
+class QuantumWhileLoop(Statement):
+    """While loops that contain only quantum statements."""
+
+    while_condition: Expression
+    block: List[QuantumStatement]
+
+
+@dataclass
+class ExternDeclaration(Statement):
+    """
+    A extern declaration
+
+    Example::
+
+        extern get_pauli(int[prec]) -> bit[2 * n];
+
+        get_pauli  // <- name
+        int[prec]  // <- classical type
+        bit[2 * n] // <- return type
+
+    """
+
+    name: str
+    classical_types: List[ClassicalType]
+    return_type: Optional[ClassicalType]
+
+
+@dataclass
+class Qubit(OpenNode):
+    """
+    A qubit
+
+    Example::
+
+        qubit q;
+
+        q  // <- Qubit
+
+    """
+
+    name: str
+
+
+class Expression(OpenNode):
+    """An expression: anything that returns a value"""
+
+
+UnaryOperator = Enum("UnaryOperator", "~ ! -")
+
+
+@dataclass
+class UnaryExpression(Expression):
+    """
+    A unary expression
+
+    Example::
+
+        ~b
+        !bool
+        -i
+
+    """
+
+    op: UnaryOperator
+    expression: Expression
+
+
+BinaryOperator = Enum("BinaryOperator", "> < >= <= == != && || | ^ & << >> + - * / % **")
```

Comment:
How difficult is it to use these binary operator enums? Another option would be to map to a name for each like `eq` and `and`

Reply:
Given the number of operators we have, it might have a better user experience if we align the symbols in the AST with the symbols in the grammar (so people don't have to constantly look it up).

Reply:
Currently, the symbols in the grammar are unnamed, l[ike this one](https://github.com/Qiskit/openqasm/blob/master/source/grammar/qasm3.g4#L445). If we split the combined grammar into lexer and parser grammar, we will have to named them in the lexer grammar.